### PR TITLE
Allow the bind network to be configured rather than inferred

### LINKED ISSUE #3689: Allow targeting IPv4 or IPv6 only
Current binding:
```
tcp6       0      0 :::4001                 :::*                    LISTEN      15408/openshift
tcp6       0      0 :::10250                :::*                    LISTEN      18556/openshift
tcp6       0      0 :::53                   :::*                    LISTEN      15408/openshift
tcp6       0      0 :::7001                 :::*                    LISTEN      15408/openshift
tcp6       0      0 :::8443                 :::*                    LISTEN      15408/openshift
```

We need to use `net.Listen("tcp4", "x.x.x.x:xxxx")` rather than the convenience `ListenAndServe` function in order to target ipv4 or ipv6

### PR DESCRIPTION
**Pull request by smarterclayton**
Allows listen on IPv4 or IPv6 only

Fixes #3689

Prevents DNS from being reached on IPv6 to prevent AAAA records from being returned (which aren't accessible).

### CODE REVIEW

**Comment by liggitt** [comment]
In validation, make sure they don't specify tcp4 with an IPv6 address or vice-versa

**Comment by smarterclayton** [comment]
Actually... that's allowed.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
If you listen on tcp4 with the automagic IPv6 address, you'll still receive incoming.  But Go will *think* you're ipv4

**Comment by liggitt** [comment]
Wha?

**Comment by smarterclayton** [comment]
Also, we can't listen on an IPv6 interface right now which I have to fix too.

**Comment by smarterclayton** [comment]
To test ipv6 listening `openshift start master --listen='https://[::1]:8443' --master="https://[::1]:8443"`.  Has some problems because endpoints don't support IPv6, but it's is a start.

**Comment by smarterclayton** [comment]
This is ready for review - OpenShift listens on IPv4 by default for the apiserver and DNS (etcd still listens on both as does the kubelet).

**Comment by smarterclayton** [comment]
I wish I could get this merged.... :(

**Comment by liggitt** [comment]
did you want to rework the skydns commit to just have ipv6/ipv4/ip in response to comments on https://github.com/skynetservices/skydns/pull/211 ?

File: pkg/cmd/flagtypes/addr.go [hunk 1]
```diff
@@ -55,26 +57,59 @@ func (a *Addr) String() string {
...
-	}
+		addr = parsed
+
+	case isIPv6Host(value):
+		a.Host = value
+		a.Port = a.DefaultPort
 
-	if strings.Contains(addr.Host, ":") {
-		host, port, err := net.SplitHostPort(addr.Host)
+	case strings.Contains(value, ":"):
```
Comment (liggitt) [inline pkg/cmd/flagtypes/addr.go]: won't this error on IPv6 addresses that contain a ":", but not a port?

File: pkg/cmd/flagtypes/addr.go [hunk 2]
```diff
@@ -99,21 +132,16 @@ func (a *Addr) Set(value string) error {
...
 			}
 		}
-		a.Host = addr.Host
+		a.Host = value
 		a.Port = port
-		addr.Host = net.JoinHostPort(addr.Host, strconv.FormatInt(int64(a.Port), 10))
-	}
-
-	if !a.AllowPrefix {
-		addr.Path = ""
```
Comment (liggitt) [inline pkg/cmd/flagtypes/addr.go]: this change probably shouldn't be part of this PR

File: pkg/cmd/server/api/validation/validation.go [hunk 3]
```diff
@@ -55,6 +55,12 @@ func ValidateServingInfo(info api.ServingInfo) fielderrors.ValidationErrorList {
 	allErrs = append(allErrs, ValidateHostPort(info.BindAddress, "bindAddress")...)
 	allErrs = append(allErrs, ValidateCertInfo(info.ServerCert, false)...)
 
+	switch info.BindNetwork {
+	case "tcp", "tcp4", "tcp6":
+	default:
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("bindNetwork", info.BindNetwork, "must be 'tcp', 'tcp4', or 'tcp6'"))
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: should we make sure the bind address matches the type of network selected? is it valid to specify an IPv4 address with 'tcp6'?

File: pkg/cmd/server/start/config_test.go [hunk 4]
```diff
@@ -76,19 +76,6 @@ func TestAssetPublicAddressDefaulting(t *testing.T) {
 	}
 }
 
-func TestAssetBindAddressDefaulting(t *testing.T) {
```
Comment (liggitt) [inline pkg/cmd/server/start/config_test.go]: why isn't this test valid any more?

Comment (liggitt) [inline pkg/cmd/server/start/config_test.go]: nevermind... forgot this became a no-op when the console unified with the master

File: pkg/cmd/server/start/node_args.go [hunk 5]
```diff
@@ -138,12 +138,7 @@ func (args NodeArgs) BuildSerializeableNodeConfig() (*configapi.NodeConfig, erro
 		config.ServingInfo.ClientCA = admin.DefaultKubeletClientCAFile(args.MasterCertDir)
 	}
 
-	// Roundtrip the config to v1 and back to ensure proper defaults are set.
```
Comment (liggitt) [inline pkg/cmd/server/start/node_args.go]: does this need to set `BindNetwork` as well? do we support it on the kubelet? do we need to?

**Comment by smarterclayton** [comment]
Want to?  No.  Will I?  Yes.

On Aug 12, 2015, at 10:34 AM, Jordan Liggitt <notifications@github.com>
wrote:

did you want to rework the skydns commit to just have ipv6/ipv4/ip in
response to comments on skynetservices/skydns#211
<https://github.com/skynetservices/skydns/pull/211> ?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3912#issuecomment-130323170>.

File: pkg/cmd/server/start/node_args.go (hunk 5 above)
Comment (liggitt) [inline pkg/cmd/server/start/node_args.go]: same question applies to NewCommandNodeConfig

File: pkg/cmd/server/api/v1/conversions.go [hunk 6]
```diff
@@ -62,6 +62,16 @@ func init() {
 				obj.ExecHandlerName = DockerExecHandlerNative
 			}
 		},
+		func(obj *ServingInfo) {
+			if len(obj.BindNetwork) == 0 {
+				obj.BindNetwork = "tcp4"
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/conversions.go]: defaulting here means this field will appear everywhere we use ServingInfo. This PR doesn't make use of it for kubelet or etcd... is that misleading?

File: pkg/cmd/util/net.go [hunk 7]
```diff
@@ -22,6 +22,69 @@ func TryListen(hostPort string) (bool, error) {
 	return true, nil
 }
 
+// tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
```
Comment (liggitt) [inline pkg/cmd/util/net.go]: maybe keep all the stuff from http/server.go in its own file with a comment on where it came from?

File: pkg/cmd/server/api/validation/validation.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/api/validation/validation.go]: ah, I see I already asked this, though I still don't understand the previous answer

File: pkg/cmd/flagtypes/addr.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/flagtypes/addr.go]: hmm... this moved? why wouldn't we enforce this in all cases

File: pkg/cmd/flagtypes/addr.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/flagtypes/addr.go]: No, isIPV6Host still returns true in that case, so you'll never get to this block if you're IPv6

File: pkg/cmd/server/api/validation/validation.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/server/api/validation/validation.go]: you can bind on tcp, tcp4, and tcp6 by providing 0.0.0.0.  All ipv4 addresses can be converted to special ipv6 addresses.

File: pkg/cmd/flagtypes/addr.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/cmd/flagtypes/addr.go]: Will look.

File: pkg/cmd/server/start/node_args.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/node_args.go]: We don't support it on the kubelet, it's not required for this commit.  This fix only requires DNS and the master to get DNS back to sanity.

File: pkg/cmd/server/api/v1/conversions.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/cmd/server/api/v1/conversions.go]: It's misleading but I don't want to have different places.  Certain config can ignore bind network now, and we can consider it a bug.

File: pkg/cmd/util/net.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/cmd/util/net.go]: It's not really hard code - I don't think it's worth it.

**Comment by smarterclayton** [comment]
I got impatient [merge]

**Comment by liggitt** [comment]
posthumous LGTM
